### PR TITLE
Print command v3

### PR DESCRIPTION
**Pull request by arnollim**
New Print Command feature prints the entire contact list, along with any insurance policies associated with each contact (whether as a beneficiary, owner or insured) into a printable .txt file of their choice. The user will name the file, for example:
print filename_1
and the file can be found in the docs/books folder as filename_1.txt

Also, code has been collated!

### CODE REVIEW

**Comment by arnollim** [comment]
Been editing a lot

File: src/main/java/seedu/address/logic/commands/PrintCommand.java [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+            + ": Saves the addressbook into a .txt file named by you for your viewing.\n"
+            + "Example: " + COMMAND_WORD + " filename\n"
+            + "file can then be found in the in doc/books folder as filename.txt";
+
+    public static final String MESSAGE_SUCCESS = "Address Book has been saved!\n"
+            + "Find your Address Book in the %1$s.txt file you created in the doc/books folder.";
+
+    private final String fileName;
+
+    public PrintCommand(String filename) {
```
Comment (Juxarius) [inline src/main/java/seedu/address/logic/commands/PrintCommand.java]: camelCase please

File: src/main/java/seedu/address/logic/commands/PrintCommand.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+                );
+                lines.add("============");
+            }
+            personIndex++;
+        }
+
+        Path file = Paths.get("docs/books/" + fileName + ".txt");
+        try {
+            Files.write(file, lines, Charset.forName("UTF-8"));
+        } catch (IOException e) {
```
Comment (Juxarius) [inline src/main/java/seedu/address/logic/commands/PrintCommand.java]: by convention we just follow the exception in naming, this should be ioe

File: src/main/java/seedu/address/logic/parser/AddressBookParser.java [hunk 3]
```diff
@@ -101,8 +102,13 @@ public Command parseCommand(String userInput) throws ParseException {
         case REDO:
             return new RedoCommand();
 
+        //@@author arnollim
```
Comment (Juxarius) [inline src/main/java/seedu/address/logic/parser/AddressBookParser.java]: code collation should be based off methods, credit the person who contributed most to the method

File: src/main/java/seedu/address/logic/parser/ParserUtil.java [hunk 4]
```diff
@@ -43,6 +48,24 @@ public static Index parseIndex(String oneBasedIndex) throws IllegalValueExceptio
...
+     */
+    public static String parseFilePath(String args) throws IllegalValueException {
+        final Matcher matcher = PRINT_ARGS_FORMAT.matcher(args.trim());
+        if (!matcher.matches()) {
+            throw new IllegalValueException(MESSAGE_INVALID_INDEX);
+        }
+        {
+            return matcher.group("filename");
+        }
+    }
```
Comment (Juxarius) [inline src/main/java/seedu/address/logic/parser/ParserUtil.java]: dont really understand the purpose of this function...? what are you parsing?

File: src/main/java/seedu/address/logic/parser/ParserUtil.java [hunk 5]
```diff
@@ -43,6 +48,24 @@ public static Index parseIndex(String oneBasedIndex) throws IllegalValueExceptio
...
+     * Leading and trailing whitespaces will be trimmed.
+     * @return "filename" for example.
+     * @throws IllegalValueException if there is no specified filepath.
+     */
+    public static String parseFilePath(String args) throws IllegalValueException {
+        final Matcher matcher = PRINT_ARGS_FORMAT.matcher(args.trim());
+        if (!matcher.matches()) {
+            throw new IllegalValueException(MESSAGE_INVALID_INDEX);
+        }
+        {
```
Comment (Juxarius) [inline src/main/java/seedu/address/logic/parser/ParserUtil.java]: what are these extra braces for? Travis not screaming at you?

File: src/main/java/seedu/address/logic/parser/PrintCommandParser.java [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+     */
+    public PrintCommand parse(String args) throws ParseException {
+        try {
+            String filename = ParserUtil.parseFilePath(args);
+            return new PrintCommand(filename);
+        } catch (IllegalValueException ive) {
+            throw new ParseException(
+                    String.format(MESSAGE_INVALID_COMMAND_FORMAT, PrintCommand.MESSAGE_USAGE));
+        }
+    }
```
Comment (Juxarius) [inline src/main/java/seedu/address/logic/parser/PrintCommandParser.java]: aaaand its just looks like you are doubling your code count now...

File: src/main/java/seedu/address/model/person/Person.java [hunk 7]
```diff
@@ -180,13 +183,25 @@ public DateOfBirth getDateOfBirth() {
         return dob.get();
     }
 
+    //@@author arnollim
     @Override
     public String getReason() {
         Address a = this.getAddress();
         Name n = this.getName();
-        this.reason = String.format(SHOWING_WHY_MESSAGE, n, a);
```
Comment (Juxarius) [inline src/main/java/seedu/address/model/person/Person.java]: Your var names are screaming at me, please go read the textbook on naming conventions, only ever use single letter for dummy vars, IDE autofills for you so why u being lazy?

File: src/main/java/seedu/address/logic/commands/WhyCommand.java [hunk 8]
```diff
@@ -35,19 +41,19 @@ public WhyCommand(Index targetIndex) {
...
-        ReadOnlyPerson personToDelete = lastShownList.get(targetIndex.getZeroBased());
-        Name name = personToDelete.getName();
-        Address address = personToDelete.getAddress();
-        String reason = personToDelete.getReason();
+
+        ReadOnlyPerson personToAnswer = lastShownList.get(targetIndex.getZeroBased());
+        Name name = personToAnswer.getName();
+        Address address = personToAnswer.getAddress();
+        String reason = personToAnswer.getReason();
         //return new CommandResult(String.format(SHOWING_WHY_MESSAGE, name, address));
```
Comment (Juxarius) [inline src/main/java/seedu/address/logic/commands/WhyCommand.java]: please also clean your code of comments, me and oscar had to help you clean up several of these. You should not be leaving grey code around if they dont give insight to people who are taking over your code

**Review by Juxarius — DISMISSED** [review]
See the related lines

File: src/main/java/seedu/address/logic/commands/PrintCommand.java (hunk 1 above)
Comment (arnollim) [reply src/main/java/seedu/address/logic/commands/PrintCommand.java]: Fixed

File: src/main/java/seedu/address/logic/commands/PrintCommand.java (hunk 2 above)
Comment (arnollim) [reply src/main/java/seedu/address/logic/commands/PrintCommand.java]: Fixed

File: src/main/java/seedu/address/logic/parser/AddressBookParser.java (hunk 3 above)
Comment (arnollim) [reply src/main/java/seedu/address/logic/parser/AddressBookParser.java]: Fixed

File: src/main/java/seedu/address/logic/parser/ParserUtil.java (hunk 5 above)
Comment (arnollim) [reply src/main/java/seedu/address/logic/parser/ParserUtil.java]: Fixed

File: src/main/java/seedu/address/model/person/Person.java (hunk 7 above)
Comment (arnollim) [reply src/main/java/seedu/address/model/person/Person.java]: fixed

File: src/main/java/seedu/address/logic/parser/ParserUtil.java (hunk 4 above)
Comment (arnollim) [reply src/main/java/seedu/address/logic/parser/ParserUtil.java]: I'm still trying to write a parser that refuses to take in the standard forbidden symbols in filenames, e.g. #./ and so on. So far I've not been successful yet, hence it seems that my parser does nothing yet

**Pull request merged**